### PR TITLE
[WIP]: DRY sampling

### PR DESCRIPTION
**Pull request by 0xymoro**
New feature (work in progress, would like some more guidance): https://github.com/vllm-project/vllm/issues/8581

Implementation of logits processor for DRY sampling. The code is a port of an optimized version of DRY that I've been running for months now across trillions of tokens and is stable. 

Some of the code was adapted from the original optimized DRY by Gemini 2.5 Pro. The logic should be the same but once we're able to hook it up and test it will see then. 

Would love some help on routing the extra_args into here so the sampler can be exposed. Got lost in the code a bit there and it's not clear to me how it works currently with old & new V1.

### CODE REVIEW

File: vllm/v1/sample/ops/penalties.py [hunk 1]
```diff
@@ -56,3 +58,202 @@ def _convert_to_tensors(output_token_ids: list[list[int]], vocab_size: int,
...
+         return logits
+
+    # Check if any request might have DRY enabled (basic check)
+    # More robust check would involve iterating through extra_data first
+    has_potential_dry = any(
+        data and data.get('dry_multiplier', _DRY_DEFAULT_MULTIPLIER) > 0
+        for data in sampling_metadata.extra_data
+    )
+    if not has_potential_dry:
+         return logits
```
Comment (NickLucche) [inline vllm/v1/sample/ops/penalties.py:110]: this check should ideally be moved to a gpu_input_batch property like penalties.

**Review by NickLucche — CHANGES_REQUESTED** [review]
Thanks for the contribution!
So the main difference is that in V1 sampling params are managed "continuously" as in they're allocated once and then overwritten as requests come through to avoid re-allocation costs.

Because of this `extra_data` field it is unclear how to handle that optimally here. Do you have more context to share on discussions about DRY as to why this could/would not be integrated like penalties are?

File: vllm/v1/sample/ops/penalties.py (hunk 1 above)
Comment (njhill) [reply vllm/v1/sample/ops/penalties.py:110]: @0xymoro apologies for the confusion. As @NickLucche says, the state that only depends on the current set of requests and their parameters, should go in `InputBatch` [here](https://github.com/vllm-project/vllm/blob/5125d72f024aa0da545e4da03ec67109353f48e4/vllm/v1/worker/gpu_input_batch.py#L187) (you can see the state associated with other sampling params there).

Then, add logic in the `add_request` method to update this state based on your own request parameters that you can retrieve from `request.sampling_params.extra_args`.

Logic should also be added to the `remove_request()`, `swap_states()` and `condense()` methods to remove/reorder the requests within the preallocated state, and to `_make_sampling_metadata()` to update the SamplingMetadata based on the current (changed) state in the input batch.

All of this is what https://github.com/vllm-project/vllm/pull/13360 aims to abstract/encapsulate into the LogitProcessor interface.

**Comment by njhill** [comment]
Thanks @0xymoro. Is this equivalent to the "new/optimized implementation" that you had mentioned in the slack thread?

Like @NickLucche says, I think there should in any case be opportunities to vectorize at least some of the operations which should make a big difference when there's a number of requests using DRY at the same time. That's mostly orthogonal to the other changes mentioned above though, I think they could be tackled in either order.

**Comment by ZeroYuJie** [comment]
#16728 Relevant processing for LogitsProcessor has been included in the current version. Will the implementation of DRY be considered in subsequent updates?

**Comment by honne23** [comment]
I would like to help push dry sampling forward.

This is particularly important for production applications when working with small LLMs (like dots.ocr). 

Ideally, integration into the standard sampling params, with good support for structured decoding. This would be a very powerful pattern.
